src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+[assembly: LoadableClass(typeof(TransformerChain<ITransformer>), typeof(TransformerChain), null, typeof(SignatureLoadModel),
+    "Transformer chain", TransformerChain.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Data
+{
+    /// <summary>
+    /// This enum allows for 'tagging' the estimators (and subsequently transformers) in the chain to be used
+    /// 'only for training', 'for training and evaluation' etc.
+    /// Most notable example is, transformations over the label column should not be used for scoring, so the scope
+    /// should be <see cref="Training"/> or <see cref="TrainTest"/>.
+    /// </summary>
+    [Flags]
+    public enum TransformerScope
+    {
+        None = 0,
+        Training = 1 << 0,
+        Testing = 1 << 1,
+        Scoring = 1 << 2,
+        TrainTest = Training | Testing,
+        Everything = Training | Testing | Scoring
+    }
+
+    /// <summary>
+    /// A chain of transformers (possibly empty) that end with a <typeparamref name="TLastTransformer"/>.
+    /// For an empty chain, <typeparamref name="TLastTransformer"/> is always <see cref="ITransformer"/>.
+    /// </summary>
+    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel, IEnumerable<ITransformer>
+    where TLastTransformer : class, ITransformer
```

Comment:
>where TLastTransformer : class [](start = 4, length = 30)

can we put it on separate line for esthetic purposes?

Reply:
I'm not sure what you meant. It's already on a separate line: `TLastTransformer` needs to be a class, and implement `ITransformer`

---
In reply to: [211688165](https://github.com/dotnet/machinelearning/pull/688#discussion_r211688165) [](ancestors = 211688165)

Reply:
oh, that ITransformer actually are requirement for TLastTransformer.
at least add one tab in front of where, otherwise my brain read this as continuation of interfaces for class, and in general across our code we always have 
private class some
      where T:
not
private class some
where T

---
In reply to: [211693191](https://github.com/dotnet/machinelearning/pull/688#discussion_r211693191) [](ancestors = 211693191,211688165)